framework/topology.cpp

```diff
@@ -1175,27 +1230,31 @@ static void init_topology_internal(const LogicalProcessorSet &enabled_cpus)
     }
 
     auto detect = [](void *ptr) -> void * {
-        const auto & enabled_cpus = *static_cast<const LogicalProcessorSet *>(ptr);
-        int curr_cpu = 0;
-        for (int i = 0; i < sApp->thread_count; ++i, ++curr_cpu) {
-            auto lp = LogicalProcessor(curr_cpu);
-            while (!enabled_cpus.is_set(lp)) {
-                lp = LogicalProcessor(++curr_cpu);
-            }
-
-            pin_to_logical_processor(lp);
-            try_detection<topo_impls>(&cpu_info[i]);
-            try_detection<ppin_impls>(&cpu_info[i]);
-            try_detection<ucode_impls>(&cpu_info[i]);
+        auto self = static_cast<TopologyDetector *>(ptr);
+        int count = sApp->thread_count;
+        [[assume(count > 0)]];
+        for (Topology::Thread &cpu : std::span(cpu_info, count)) {
+            pin_to_logical_processor(LogicalProcessor(cpu.cpu_number));
+            self->detect_via_cpuid(&cpu);
+            self->detect_via_os(&cpu);
         }
         return nullptr;
     };
 
-    pthread_t detection_thread;
-    pthread_create(&detection_thread, nullptr, detect, const_cast<LogicalProcessorSet *>(&enabled_cpus));
-    pthread_join(detection_thread, nullptr);
+    if (setup_cpuid_detection()) {
+        // Do a per-CPU detection, which requires us to pin an auxiliary
+        // thread to each CPU.
+        pthread_t detection_thread;
+        pthread_create(&detection_thread, nullptr, detect, this);
+        pthread_join(detection_thread, nullptr);
+    } else {
+        // Use OS-level detection only (no pinning, no threads).
+        for (Topology::Thread &cpu : std::span(cpu_info, count)) {
+            detect_via_os(&cpu);
+        }
+    }
 
-    fill_numa();
+    detect_numa();
```

Comment:
Maybe a corner case, but there is no definition for `detect_numa()` when OS is different from Linux and Windows.

Reply:
I haven't compiled FreeBSD yet. Thanks for noticing.